File: examples/go/boltutil_demo.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

var (
	apiBase       = env("BOLT_API_BASE", "https://api.boltutil.com")
	apiKey        = env("BOLT_API_KEY", "YOUR_API_KEY")
	webhookSecret = env("BOLT_WEBHOOK_SECRET", "YOUR_WEBHOOK_SECRET")
	notifyURL     = env("BOLT_NOTIFY_URL", "https://merchant.example.com/webhooks/boltutil")
	returnURL     = env("BOLT_RETURN_URL", "https://merchant.example.com/orders/return")
)

func env(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}

func hmacSHA256Hex(payload, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(payload))
	return hex.EncodeToString(mac.Sum(nil))
}

func signBody(rawBody, timestamp string) string {
	return hmacSHA256Hex(timestamp+"."+rawBody, webhookSecret)
}

func createOrder(network string) error {
	externalOrderID := fmt.Sprintf("INV-%d", time.Now().UnixMilli())
	payload := map[string]any{
		"amount":          25.0,
		"externalOrderId": externalOrderID,
		"network":         network,
		"currency":        "USDT",
		"orderDesc":       "BoltUtil demo order",
		"notifyUrl":       notifyURL,
		"returnUrl":       returnURL + "?order=" + externalOrderID,
		"expiredMinutes":  30,
		"metadata": map[string]any{
			"customerId": "CUS_1001",
			"source":     "go-demo",
		},
	}

	bodyBytes, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	rawBody := string(bodyBytes)
	timestamp := strconv.FormatInt(time.Now().UnixMilli(), 10)
	signature := signBody(rawBody, timestamp)

	req, err := http.NewRequest("POST", apiBase+"/api/v1/order/create", bytes.NewBufferString(rawBody))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Bolt-Key", apiKey)
	req.Header.Set("X-Bolt-Timestamp", timestamp)
	req.Header.Set("X-Bolt-Signature", signature)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	responseBody, _ := io.ReadAll(resp.Body)
	fmt.Println("HTTP", resp.StatusCode)
	fmt.Println(string(responseBody))
	return nil
}

func verifyWebhook(rawBody, timestamp, signature string) bool {
	if timestamp == "" || signature == "" {
		return false
	}

	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return false
	}
	if abs(time.Now().UnixMilli()-ts) > 5*60*1000 {
		return false
	}

	expected := signBody(rawBody, timestamp)
	left := []byte(strings.ToLower(expected))
	right := []byte(strings.ToLower(signature))
	return len(left) == len(right) && subtle.ConstantTimeCompare(left, right) == 1
}

func abs(value int64) int64 {
	if value < 0 {
		return -value
	}
	return value
}

func webhookHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.NotFound(w, r)
		return
	}

	bodyBytes, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, `{"error":"read_body_failed"}`, http.StatusBadRequest)
		return
	}
	rawBody := string(bodyBytes)
	timestamp := r.Header.Get("X-Bolt-Webhook-Timestamp")
	signature := r.Header.Get("X-Bolt-Webhook-Signature")

	w.Header().Set("Content-Type", "application/json")
	if !verifyWebhook(rawBody, timestamp, signature) {
		w.WriteHeader(http.StatusUnauthorized)
		w.Write([]byte(`{"error":"invalid_signature"}`))
		return
	}

	var event map[string]any
	if err := json.Unmarshal(bodyBytes, &event); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(`{"error":"invalid_json"}`))
		return
	}

	fmt.Println("Verified BoltUtil webhook:", event)
	// TODO: update your local order by event["externalOrderId"].
	// Only fulfill when status is CONFIRMED or COMPLETED, according to your business rules.

	w.Write([]byte(`{"status":"SUCCESS"}`))
}

func startWebhookServer() error {
	port := env("PORT", "3000")
	http.HandleFunc("/webhooks/boltutil", webhookHandler)
	fmt.Println("BoltUtil webhook demo listening on http://localhost:" + port + "/webhooks/boltutil")
	return http.ListenAndServe(":"+port, nil)
}

func main() {
	command := ""
	if len(os.Args) > 1 {
		command = os.Args[1]
	}

	switch command {
	case "create":
		network := "TRC20"
		if len(os.Args) > 2 {
			network = os.Args[2]
		}
		if err := createOrder(network); err != nil {
			panic(err)
		}
	case "webhook":
		if err := startWebhookServer(); err != nil {
			panic(err)
		}
	default:
		fmt.Println("Usage:")
		fmt.Println("  go run boltutil_demo.go create TRC20")
		fmt.Println("  go run boltutil_demo.go create ERC20")
		fmt.Println("  go run boltutil_demo.go create BEP20")
		fmt.Println("  go run boltutil_demo.go create POLYGON")
		fmt.Println("  go run boltutil_demo.go create SOLANA")
		fmt.Println("  go run boltutil_demo.go webhook")
	}
}

File: examples/python/boltutil_demo.py
import hashlib
import hmac
import json
import os
import sys
import time
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.request import Request, urlopen
from urllib.error import HTTPError

API_BASE = os.getenv("BOLT_API_BASE", "https://api.boltutil.com")
API_KEY = os.getenv("BOLT_API_KEY", "YOUR_API_KEY")
WEBHOOK_SECRET = os.getenv("BOLT_WEBHOOK_SECRET", "YOUR_WEBHOOK_SECRET")
NOTIFY_URL = os.getenv("BOLT_NOTIFY_URL", "https://merchant.example.com/webhooks/boltutil")
RETURN_URL = os.getenv("BOLT_RETURN_URL", "https://merchant.example.com/orders/return")


def hmac_sha256_hex(payload: str, secret: str) -> str:
    return hmac.new(secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256).hexdigest()


def sign_body(raw_body: str, timestamp: str) -> str:
    return hmac_sha256_hex(f"{timestamp}.{raw_body}", WEBHOOK_SECRET)


def create_order(network: str = "TRC20") -> None:
    external_order_id = f"INV-{int(time.time() * 1000)}"
    payload = {
        "amount": 25.0,
        "externalOrderId": external_order_id,
        "network": network,
        "currency": "USDT",
        "orderDesc": "BoltUtil demo order",
        "notifyUrl": NOTIFY_URL,
        "returnUrl": f"{RETURN_URL}?order={external_order_id}",
        "expiredMinutes": 30,
        "metadata": {
            "customerId": "CUS_1001",
            "source": "python-demo",
        },
    }
    raw_body = json.dumps(payload, separators=(",", ":"), ensure_ascii=False)
    timestamp = str(int(time.time() * 1000))
    signature = sign_body(raw_body, timestamp)

    request = Request(
        f"{API_BASE}/api/v1/order/create",
        data=raw_body.encode("utf-8"),
        method="POST",
        headers={
            "Content-Type": "application/json",
            "X-Bolt-Key": API_KEY,
            "X-Bolt-Timestamp": timestamp,
            "X-Bolt-Signature": signature,
        },
    )

    try:
        with urlopen(request, timeout=20) as response:
            print("HTTP", response.status)
            print(response.read().decode("utf-8"))
    except HTTPError as error:
        print("HTTP", error.code)
        print(error.read().decode("utf-8"))


def verify_webhook(raw_body: str, timestamp: str, signature: str) -> bool:
    if not timestamp or not signature:
        return False
    if abs(int(time.time() * 1000) - int(timestamp)) > 5 * 60 * 1000:
        return False
    expected = sign_body(raw_body, timestamp)
    return hmac.compare_digest(expected.lower(), signature.lower())


class BoltUtilWebhookHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/webhooks/boltutil":
            self.send_response(404)
            self.end_headers()
            return

        length = int(self.headers.get("Content-Length", "0"))
        raw_body = self.rfile.read(length).decode("utf-8")
        timestamp = self.headers.get("X-Bolt-Webhook-Timestamp", "")
        signature = self.headers.get("X-Bolt-Webhook-Signature", "")

        if not verify_webhook(raw_body, timestamp, signature):
            self.send_response(401)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(b'{"error":"invalid_signature"}')
            return

        event = json.loads(raw_body)
        print("Verified BoltUtil webhook:", event)

        # TODO: update your local order by event["externalOrderId"].
        # Only fulfill when status is CONFIRMED or COMPLETED, according to your business rules.

        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(b'{"status":"SUCCESS"}')


def start_webhook_server(port: int = 3000) -> None:
    server = HTTPServer(("0.0.0.0", port), BoltUtilWebhookHandler)
    print(f"BoltUtil webhook demo listening on http://localhost:{port}/webhooks/boltutil")
    server.serve_forever()


if __name__ == "__main__":
    command = sys.argv[1] if len(sys.argv) > 1 else ""
    if command == "create":
        create_order(sys.argv[2] if len(sys.argv) > 2 else "TRC20")
    elif command == "webhook":
        start_webhook_server(int(os.getenv("PORT", "3000")))
    else:
        print("Usage:")
        print("  python3 boltutil_demo.py create TRC20")
        print("  python3 boltutil_demo.py create ERC20")
        print("  python3 boltutil_demo.py create BEP20")
        print("  python3 boltutil_demo.py create POLYGON")
        print("  python3 boltutil_demo.py create SOLANA")
        print("  python3 boltutil_demo.py webhook")
